### PR TITLE
Messaging interface

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: message-viewer-directive.js [hunk 1]
```diff
@@ -20,35 +20,74 @@ function factory(brMessagesService, $routeParams) {
     var model = scope.model = {};
     model.message = null;
 
+    model.foobar = 'hey';
```
Comment (mattcollier) [inline message-viewer-directive.js]: remove.

File: message-viewer-directive.js [hunk 2]
```diff
@@ -20,35 +20,74 @@ function factory(brMessagesService, $routeParams) {
...
+        }
+        $location.url(brMessagesService.messageListUrl);
       }
     };
 
+    model.delete = function() {
+      //TODO need delete API
+      brMessagesService.delete(model.message)
+        .then(function(response) {
+          console.log('Delete success');
```
Comment (mattcollier) [inline message-viewer-directive.js]: remove console logging throughout.

File: message-viewer-directive.js [hunk 3]
```diff
@@ -20,35 +20,74 @@ function factory(brMessagesService, $routeParams) {
...
+          // Mark selected objects as archived
+          model.viewMessages();
+        })
+        .catch(function(response) {
```
Comment (mattcollier) [inline message-viewer-directive.js]: remove logging, is this error adequately dealt with?

File: messages-list-directive.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ define([], function() {
 'use strict';
 
 /* @ngInject */
-function factory(brMessagesService, $location) {
+function factory(brMessagesService, $location, $http, config) {
```
Comment (mattcollier) [inline messages-list-directive.js]: remove $http and config if not needed.

File: messages-list-directive.js [hunk 5]
```diff
@@ -21,40 +21,145 @@ function factory(brMessagesService, $location) {
...
-            model.sorting.date + 'content.date'
-          ];
-          break;
         case 'date':
           model.sorting.date = (model.sorting.date === '+') ? '-' : '+';
           model.orderBy = [
-            model.sorting.date + 'content.date',
-            model.sorting.subject + '_id'
+            model.sorting.date + 'date',
+            model.sorting.subject + 'id'
```
Comment (mattcollier) [inline messages-list-directive.js]: is this still applicable, if so, probably shouldn't be 'id' anymore.

File: messages-service.js [hunk 6]
```diff
@@ -28,13 +31,56 @@ function factory($http, config) {
...
-            return !message.meta.read;
+            return !message.meta.read && !message.meta.archived;
           }).length;
           return new Promise(function(resolve, reject) {
             resolve(results);
           });
         });
   };
+
+  service.delete = function(message) {
```
Comment (mattcollier) [inline messages-service.js]: Wrap this in Promise as with service.get.

File: messages-service.js [hunk 7]
```diff
@@ -28,13 +31,56 @@ function factory($http, config) {
...
+    return $http.delete(messagesEndpoint + '/' + message.id);
+  };
+
+  service.deleteBatch = function(messages) {
+    var request = {};
+    request.messages = messages;
+
+    var dataString = JSON.stringify(request);
+
+    return $http({
```
Comment (mattcollier) [inline messages-service.js]: Wrap this in Promise as with service.get.

File: messages-service.js [hunk 8]
```diff
@@ -28,13 +31,56 @@ function factory($http, config) {
...
+  };
+
+  service.archive = function(message) {
+    var request = {};
+    request.operation = 'archive';
+    request.message = message.id;
+
+    var dataString = JSON.stringify(request);
+
+    return $http.post(messagesEndpoint + '/' + message.id, dataString);
```
Comment (mattcollier) [inline messages-service.js]: Wrap this in Promise as with service.get.

File: messages-service.js [hunk 9]
```diff
@@ -28,13 +31,56 @@ function factory($http, config) {
...
+  };
+
+  service.archiveBatch = function(messages) {
+    var request = {};
+    request.operation = 'archive';
+    request.messages = messages;
+
+    var dataString = JSON.stringify(request);
+
+    return $http.post(messagesBatchEndpoint, dataString);
```
Comment (mattcollier) [inline messages-service.js]: Wrap this in Promise as with service.get.

**Comment by dlongley** [comment]
Assigning this to you @lamara to let me know if this PR should be merged or not -- or if you're still doing some reorganization.

**Comment by lamara** [comment]
This can be merged.

File: message-viewer-directive.js [hunk 10]
```diff
@@ -19,36 +19,71 @@ function factory(brMessagesService, $routeParams) {
...
+        $location.url(brMessagesService.messageListUrl);
       }
     };
 
+    model.delete = function() {
+      brMessagesService.delete(model.message)
+        .then(function(result) {
+          model.viewMessages();
+          scope.$apply();
+        }, function(error) {
```
Comment (dlongley) [inline message-viewer-directive.js]: Use `.catch(function(err) { ...` syntax, its semantics are clearer and therefore, it's more readable.

File: message-viewer-directive.js [hunk 11]
```diff
@@ -19,36 +19,71 @@ function factory(brMessagesService, $routeParams) {
...
+          scope.$apply();
+        });
+    };
+
+    model.archive = function() {
+      brMessagesService.archive(model.message)
+        .then(function(result) {
+          model.viewMessages();
+          scope.$apply();
+        }, function(error) {
```
Comment (dlongley) [inline message-viewer-directive.js]: Use `.catch(function(err) { ...` syntax, its semantics are clearer and therefore, it's more readable.

File: message-viewer.html [hunk 12]
```diff
@@ -1,5 +1,77 @@
-<div class="row">
-  <div class="col-sm-offset-2 col-sm-8">
-    {{model.message | json}}
+<div ng-if="model.message" class="row">
+  <div class="row">
+  	<h1 class="headline">View Messages</h1>
```
Comment (dlongley) [inline message-viewer.html]: 2 space indent.

File: message-viewer.html [hunk 13]
```diff
@@ -1,5 +1,77 @@
...
+  </div>
+  <div class="row">
+    <h4 ng-if="model.loading" class="col-sm-offset-2 col-sm-8">Loading <i class="fa fa-refresh fa-spin text-muted"></i></h4>
+  </div>
+  <div ng-show="!model.loading" class="col-sm-offset-2 col-sm-8">
+    <div class="row alert alert-danger" ng-show="model.showFailAlert">
+      <strong ng-if="!model.message.meta.archived">Error: Message could not be archived</strong>
+      <strong ng-if="model.message.meta.archived">Error: Message could not be deleted</strong>
+    </div>
+  	<div class="row" style="border-bottom: 1px solid #666; margin-bottom: 20px;">
```
Comment (dlongley) [inline message-viewer.html]: Indent off throughout this file.

File: message-viewer.html [hunk 14]
```diff
@@ -1,5 +1,77 @@
...
+<div ng-show="!model.loading" ng-if="model.message == null" class="row">
+  <div class="row">
+    <h1 class="headline">Error retrieving message</h1>
+  </div>
+  <div class="row">
+    <div class="col-sm-offset-5 col-sm-3">
+      <button style="margin-top: 40px;" type="button" class="btn btn-primary" ng-click="model.viewMessages()" style="" ng-click="">Return To Messages</button>
+    </div>
+  </div>
+</div>
```
Comment (dlongley) [inline message-viewer.html]: Add EOL.

File: messages-list-directive.js [hunk 15]
```diff
@@ -19,42 +19,148 @@ function factory(brMessagesService, $location) {
...
+      }
+      processMessageLists();
+    };
+
+    model.archiveTab = function() {
+      model.inboxActive = false;
+      model.archiveActive = true;
+      model.showDeleteSuccessAlert = false;
+      model.showDeleteFailAlert = false;
+      for(var message in allMessages) {
```
Comment (dlongley) [inline messages-list-directive.js]: Don't use `for(var x in y)` for arrays. Use `forEach` or `for(var i = 0; i < x.length; ++i) {`.

File: messages-list-directive.js [hunk 16]
```diff
@@ -19,42 +19,148 @@ function factory(brMessagesService, $location) {
...
+        .then(function(result) {
+          // Mark selected objects as archived
+          selectedMessages.forEach(function(message) {
+            message.meta.archived = true;
+          });
+          model.showDeleteSuccessAlert = true;
+          model.showDeleteFailAlert = false;
+          processMessageLists();
+          scope.$apply();
+        }, function(error) {
```
Comment (dlongley) [inline messages-list-directive.js]: Use `catch(function(err) {...`.

File: message-viewer-directive.js (hunk 11 above)
Comment (lamara) [inline message-viewer-directive.js]: I remember trying to use .catch syntax, but it didn't play well with the promise returned by the message service (we wrap the http calls in a Promise.resolve() -- I noticed that we could use .catch syntax when we didn't wrap it, but @mattcollier asked to wrap all of our calls so .catch stopped working). Any idea why that's happening?

File: messages-service.js [hunk 17]
```diff
@@ -28,13 +31,59 @@ function factory($http, config) {
...
+
+  service.delete = function(message) {
+    return Promise.resolve($http.delete(messagesEndpoint + '/' + message.id));
+  };
+
+  service.deleteBatch = function(messages) {
+    var request = {};
+    request.messages = messages;
+
+    var dataString = JSON.stringify(request);
```
Comment (dlongley) [inline messages-service.js]: Could just move this `JSON.stringify` down and remove the `dataString` var for brevity.

File: messages-list-directive.js [hunk 18]
```diff
@@ -19,42 +19,148 @@ function factory(brMessagesService, $location) {
...
+    processMessageLists();
+
+    // Update the message service with the list's current URL so we can path back to it later
+    brMessagesService.setMessageListUrl($location.path());
 
     brMessagesService.getAll()
       .then(function(result) {
-        model.messages = result.data;
+        allMessages = result.data;
+        for(var message in allMessages) {
```
Comment (dlongley) [inline messages-list-directive.js]: Don't use `for x in y` with arrays. Use `forEach` or `for(var i = 0; i < allMessages.length; ++i)`

File: messages-list-directive.js [hunk 19]
```diff
@@ -19,42 +19,148 @@ function factory(brMessagesService, $location) {
...
           break;
       }
     };
 
+    model.inboxTab = function() {
+      model.inboxActive = true;
+      model.archiveActive = false;
+      model.showDeleteSuccessAlert = false;
+      model.showDeleteFailAlert = false;
+      for(var message in allMessages) {
```
Comment (dlongley) [inline messages-list-directive.js]: Use `forEach` or for w/a counter, not `for x in y`.

File: message-viewer-directive.js (hunk 11 above)
Comment (mattcollier) [inline message-viewer-directive.js]: Here's an example of .catch being used with the Promise polyfill: https://github.com/digitalbazaar/bedrock-angular-key/blob/master/generate-key-pair-modal-directive.js#L44-L78

**Comment by dlongley** [comment]
Reassigned to @lamar with the assumption that some changes will be required here due to changes to the HTTP API.

**Pull request merged**